[asr->fortran] implement visit_{FileOpen, FileClose, FileRead}

khushi-411 opened the pull request:
#### TODO
- [x]  enable remaining valid tests

Thirumalai-Shaktivel commented on src/libasr/codegen/asr_to_fortran.cpp:
```diff
@@ -558,11 +558,100 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
+            visit_expr(*x.m_newunit);
+            r += s;
+        } else {
+            r += "*, ";
```
```suggestion
            r += "*";
```

Thirumalai-Shaktivel commented on src/libasr/codegen/asr_to_fortran.cpp:
```diff
@@ -558,11 +558,100 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
+            visit_expr(*x.m_unit);
+            r += s;
+        } else {
+            r += "*, ";
```
```suggestion
            r += "*";
```

Thirumalai-Shaktivel commented on src/libasr/codegen/asr_to_fortran.cpp:
```diff
@@ -558,11 +558,100 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
+        if (x.m_unit) {
+            visit_expr(*x.m_unit);
+            r += s;
+        }
```
```suggestion
        } else {
            r += "*";
        }
```

Thirumalai-Shaktivel reviewed (commented):
It looks good so far!
If you feel you have finished the PR, please mark it as "Ready for Review".

khushi-411 commented:
Addressed all your reviews! Thanks, @Thirumalai-Shaktivel!

> If you feel you have finished the PR, please mark it as "Ready for Review".

I think we can also enable some other tests too. For example: `file_04.f90`, these tests pass along with the `NOFAST` flag but not without it. This is the generated code for it:

<details>
<summary>generated code</summary>

<p>

```f90
program load_model_program
implicit none
integer(4) :: int_data1
integer(4) :: int_data2
real(4) :: real_data1
real(4) :: real_data2
integer(4), save :: u = 11
open(u, file="file_04_data.dat", form="unformatted")
read(u, *) int_data1, real_data1, int_data2, real_data2
print *, "Data: ", int_data1, real_data1, int_data2, real_data2
if (int_data1 /= 345239) then
    error stop
end if
if (abs(real_data1 - 52.910000) > 0.000001) then
    error stop
end if
if (int_data2 /= (-345239)) then
    error stop
end if
if (abs(real_data2 - (-52.910000)) > 0.000001) then
    error stop
end if
end program load_model_program
```

</p>
</details>

<details>
<summary>test code</summary>

<p>

```f90
program load_model_program
    implicit none

    integer :: int_data1, int_data2
    real :: real_data1, real_data2
    integer :: u = 11

    open(u, file="file_04_data.dat", form="unformatted", access="stream")
    read(u) int_data1, real_data1, int_data2, real_data2

    print *, "Data: ", int_data1, real_data1, int_data2, real_data2
    if (int_data1 /= 345239) error stop
    if (abs(real_data1 - (52.91)) > 1e-6) error stop
    if (int_data2 /= -345239) error stop
    if (abs(real_data2 - (-52.91)) > 1e-6) error stop
end program

```

</p>
</details>

What do you think?

Shaikh-Ubaid commented on integration_tests/CMakeLists.txt:
```diff
@@ -870,19 +870,19 @@ RUN(NAME return_04 LABELS gfortran llvm wasm)
-RUN(NAME file_05       LABELS gfortran llvm COPY_TO_BIN file_05_data.dat)
-RUN(NAME file_06       LABELS gfortran llvm COPY_TO_BIN file_06_data.dat)
+RUN(NAME file_03       LABELS gfortran llvm fortran COPY_TO_BIN file_03_data.txt)
+RUN(NAME file_04       LABELS gfortran llvm fortran NOFAST COPY_TO_BIN file_04_data.dat)
+RUN(NAME file_05       LABELS gfortran llvm fortran NOFAST COPY_TO_BIN file_05_data.dat)
+RUN(NAME file_06       LABELS gfortran llvm fortran NOFAST COPY_TO_BIN file_06_data.dat)
```
Why is the NOFAST added here? By adding NOFAST we are removing the --fast versions of these tests. We always try to support all the available integration tests and in the rarest of the rare cases might allow removing a test to be supported later.

Is there a way to add support in the fortran backend so that it works for the above tests without the addition of NOFAST?

certik replied:
Yes, NOFAST should not be added.

Thirumalai-Shaktivel commented on src/libasr/codegen/asr_to_fortran.cpp:
```diff
@@ -580,11 +580,102 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
+            visit_expr(*x.m_unit);
+            r += s;
+        } else {
+            r += "*";
```
I haven't seen a case `close(*)`, but it's fine.

Thirumalai-Shaktivel replied:
I went through Fortran 2018 document, pg. 223.
The syntax doesn't make sense. So, `close(*)` is invalid.
I think we can remove it.

khushi-411 replied:
Thanks for the informative share, @Thirumalai-Shaktivel! I wasn't aware of that document.

Have updated the PR. Thanks!

Shaikh-Ubaid replied:
Good point, Thirumalai! Since `open()` and `close()` must have a unit specified, how about we throw a `CodeGenError()` in the else case when the unit is `nullptr`?

Thirumalai-Shaktivel reviewed (approved):
LGTM, Thank you!

Thirumalai-Shaktivel commented:
Thank you!

Thirumalai-Shaktivel commented:
@Shaikh-Ubaid, can you review and merge this PR?

Shaikh-Ubaid reviewed (approved):
I shared a minor comment. This looks good to me! Very valuable contributions! Thanks for this!

khushi-411 commented:
Thanks for all your helpful reviews in the PR, @Thirumalai-Shaktivel & @Shaikh-Ubaid!

Have addressed all of your reviews!
